docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,241 @@
+| ------------- | ------------- | ------------- |
+| `getmetatable()` | `typelib` | returns the runtime representation of self's `Metatable` |
+
+#### StringSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `string` | returns self's value of a string singleton |
+
+#### BooleanSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `boolean` | returns self's boolean singleton value of either `true` or `false` |
+
+#### Table
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `addprops(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |
+| `delprops(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it; if the key doesn't exist, nothing happens |
+| `getprops()` | `{typelib: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{typelib.getstringsingleton("age") = typelib.getnumber()}`) |
+| `settableindexer(indexty: typelib, indexresultty: typelib)` | `nil` | sets self's index type to the first argument and index result type to the second |
+| `getindextype()` | `typelib` | returns self's index type |
+| `getindexresulttype()` | `typelib` | returns self's index result type |
+| `setsealed(arg: boolean)` | `nil` | sets the table's sealed status. If true, the table becomes a sealed table. Default initialization of tables is unsealed tables (aka setsealed(false)) |
+| `issealed()` | `boolean` | returns the seal status of self |
+
+#### Metatable
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `settable(arg: typelib)` | `nil` | sets self's table to the argument; the argument needs to be `istable()` |
+| `gettable()` | `typelib` | returns self's runtime representation of the type `table` |
+| `setmetatable(arg: typelib)` | `nil` | sets self's metatable to the argument; the argument needs to be `ismetatable()` |
+| `getmetatable()` | `typelib` | returns self's runtime representation of `Metatable` |
+
+#### Function
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `setargtypes(arg: {typelib} \| typelib)` | `nil` | sets self's argument types to the argument, where an array implies a TypePack and the latter implies a Variadic |
+| `getargtypes()` | `{typelib} \| typelib` | returns the runtime representation of self's argument type |
+| `setreturntypes(arg: {typelib} \| typelib)` | `nil` | sets self's return types to the argument, where an array implies a TypePack and the latter implies a Variadic |
+| `getreturntypes()` | `{typelib} \| typelib` | returns the runtime representation of self's return type |
```

Comment:
There's a footgun here: rank-N. If you create a type parameter and you wish to create a function type with that type parameter, then the type parameter will not be instantiated when the function is called.

You would need `addtypeparam`, but now a new footgun: what if you forgot to call it.

Reply:
Another footgun: calling `addtypeparam` on two functions to form `<a>(a) -> <a>(a) -> a` where `<a>` in both function types here are the _same_ type parameter.